_posts/2024-08-14-running-a-hsm-realm.md

```diff
@@ -0,0 +1,221 @@
+randomize the hash used to generate a merkle tree record id from the tenant and
+user information. If you have the Google `gcloud` tools installed then this
+snippet can be used to create the key.
+
+```sh
+python3 -c \
+'import secrets, json; print(json.dumps({"data": secrets.token_hex(32), "encoding": "Hex", "algorithm": "Blake2sMac256"}), end="")' | \
+    gcloud \
+        --project <gcp-project> \
+        secrets create \
+        record-id-randomization \
+        --data-file - \
+        --labels kind=record_id_randomization_key
+```
+
+## Realm Initialization
+
+Once the agent is successfully running on each HSM server the realm
+initialization can be performed. The
+[`cluster`](https://github.com/juicebox-systems/juicebox-hsm-realm/tree/main/cluster_cli)
+tool is used to perform many administrative tasks. The `cluster agents` command
+can be used to sanity check that the agents are running and accessible.
+
+First use `cluster new-realm <some-agent>` to crete a new realm. This will
+assign the realm its IDs and create and persist the initial merkle tree.
+
+Then use `cluster join-realm --realm <realmID> <agent1> <agent2> ..` to have the
+other HSMs join the same realm.
+
+The initial realm creation created a single replication group with a single
+member. At this point new replication groups can be created (via `cluster
+new-group`) and the groups given ownership of part of the key space (via
+`cluster transfer`). The `cluster groups` command can give a detailed report of
+each replication groups status.
+
+You'll need to decide the exact number of replication groups, and the
+distribution of their members. Differing sizes of clusters may use different
+approaches. For a 5 HSM cluster we recommend 3 replication groups each
+containing all 5 HSMs as members and each replication group owning a third of
+the key space. In larger clusters its not required that all HSMs are members of
+all replication groups.
+
+## Services
+
+The agent, load balancer and cluster manager services will all panic in
+unexpected and unrecoverable error conditions. A process supervisor such as
+Systemd should be used to manage these services and restart the process as
+needed.
+
+The services generate logs to `stdout` and publish metrics to Datadog if the
+Datadog agent is installed.
+
+The `service_checker` tool can be used to repeated make requests to a realm and
+report errors via a Datadog service-check.
+
+### Service Discovery
+
+Services in the realm use service discovery to find other instances. Service
+discovery data is stored in Bigtable. In order to start a service an IP and port
+to bind to needs to be provided via the command line arguments. This IP should
+be reachable from other instances in the realm. i.e. It should not be localhost
+or 0.0.0.0.
+
+## Tenants
+
+Requests to the realm authenticate using JWT. The JWT key(s) for a tenant are
+store in Secret Manager. The `gcloud` tool can be useful in creating these, e.g.
```

Comment:
Maybe should also link to the `tokens` CLI tool in the SDK repo for creating more complex tokens, like RSA and 25519?